phoenix-core-server/src/main/java/org/apache/phoenix/replication/log/LogFileCodec.java

```diff
@@ -0,0 +1,288 @@
+
+            Mutation mutation = record.getMutation();
+            LogFileRecord.MutationType mutationType = LogFileRecord.MutationType.get(mutation);
+            recordOut.writeByte(mutationType.getCode());
+            byte[] nameBytes = record.getHBaseTableName().getBytes(StandardCharsets.UTF_8);
+            WritableUtils.writeVInt(recordOut, nameBytes.length);
+            recordOut.write(nameBytes);
+            WritableUtils.writeVLong(recordOut, record.getCommitId());
+            byte[] rowKey = mutation.getRow();
+            WritableUtils.writeVInt(recordOut, rowKey.length);
+            recordOut.write(rowKey);
+            recordOut.writeLong(mutation.getTimestamp());
+
+            Map<byte[], List<Cell>> familyMap = mutation.getFamilyCellMap();
+            int cfCount = familyMap.size();
+            WritableUtils.writeVInt(recordOut, cfCount);
+
+            for (Map.Entry<byte[], List<Cell>> entry: familyMap.entrySet()) {
+                byte[] columnFamily = entry.getKey();
+                WritableUtils.writeVInt(recordOut, columnFamily.length);
+                recordOut.write(columnFamily);
+                List<Cell> cells = entry.getValue();
+                WritableUtils.writeVInt(recordOut, cells.size());
+                for (Cell cell: cells) {
+                    WritableUtils.writeVInt(recordOut, cell.getQualifierLength());
+                    recordOut.write(cell.getQualifierArray(), cell.getQualifierOffset(),
+                        cell.getQualifierLength());
+                    WritableUtils.writeVInt(recordOut, cell.getValueLength());
+                    if (cell.getValueLength() > 0) {
+                        recordOut.write(cell.getValueArray(), cell.getValueOffset(),
+                            cell.getValueLength());
+                    }
+                }
+            }
+
+            byte[] currentRecordBytes = currentRecord.toByteArray();
+            // Write total record length
+            WritableUtils.writeVInt(out, currentRecordBytes.length);
+            // Write the record
+            out.write(currentRecordBytes);
+
+            // Set the size (including the vint prefix) on the record object
+            ((LogFileRecord) record).setSerializedLength(currentRecordBytes.length
+                + WritableUtils.getVIntSize(currentRecordBytes.length));
+
+            // Reset the ByteArrayOutputStream to release resources
+            currentRecord.reset();
+        }
+    }
+
+    private static class RecordDecoder implements LogFile.Codec.Decoder {
+        private final DataInput in;
+        // A reference to the object populated by the last successful advance()
+        private LogFileRecord current = null;
+
+        RecordDecoder(DataInput in) {
+            this.in = in;
+        }
+
+        @Override
+        public boolean advance(LogFile.Record reuse) throws IOException {
+            try {
+                int recordDataLength = WritableUtils.readVInt(in);
```

Comment:
Instead of using an EOFException to determine the end of a block, should we keep track of the position within a block to improve CPU utilization?

Reply:
How would that improve CPU utilization?

Reply:
I could not see the check for the end of the current block. Where is it?

Reply:
This is the Codec, block level concepts are not available here, like current block start and end offsets. A codec is given a stream and it reads from that stream a structure that is self describing in the sense that internally any field has a length preceding it. At any point, any read may throw EOFException though, that cannot be removed.

Reply:
I have looked there too. There is a check for the end of the data offset. That is essentially checking the end of file essentially. There we read the block header and payload. If the payload is compressed, we decompressed it. The block now in memory as a `ByteBuffer`. We know the buffer size and can check if the end of buffer is reached before we call `advance`. However, we do not do this check.